File: internal/application/queries/list_audits.go
package queries

import (
	"context"
	"fmt"

	"github.com/memlore/memlore/internal/application/ports"
	"github.com/memlore/memlore/internal/domain"
)

// ListAuditsHandler lists audit records for a lore entry.
type ListAuditsHandler struct {
	begin ports.UnitOfWorkFactory
}

func NewListAuditsHandler(begin ports.UnitOfWorkFactory) *ListAuditsHandler {
	return &ListAuditsHandler{begin: begin}
}

func (h *ListAuditsHandler) Handle(ctx context.Context, entryID string) ([]domain.AuditRecord, error) {
	uow, err := h.begin(ctx)
	if err != nil {
		return nil, err
	}
	defer uow.Rollback(ctx)

	if _, err := uow.LoreEntries().Get(ctx, entryID); err != nil {
		if _, ok := err.(*domain.NotFoundError); ok {
			return nil, &domain.NotFoundError{
				Message: fmt.Sprintf("lore entry %s not found", entryID),
			}
		}
		return nil, err
	}
	return uow.Audits().ListByTarget(ctx, entryID)
}

File: src/memlore/application/queries/list_audits.py
from __future__ import annotations

from memlore.application.ports.unit_of_work import UnitOfWork
from memlore.domain.exceptions import NotFoundError
from memlore.domain.models.audit_record import AuditRecord


class ListAuditsHandler:
    def __init__(self, uow: UnitOfWork) -> None:
        self._uow = uow

    def handle(self, entry_id: str) -> list[AuditRecord]:
        if self._uow.lore_entries.get(entry_id) is None:
            raise NotFoundError(f"lore entry {entry_id} not found")
        return self._uow.audits.list_by_target(entry_id)
